balancer-js/src/modules/vaultModel/poolModel/join.ts

```diff
@@ -0,0 +1,245 @@
+      const bptAmountOut =
+        WeightedPoolDecoder.joinAllTokensInForExactBPTOut(encodedUserData);
+      return bptAmountOut.toString();
+    } else if (joinKind === WeightedPoolJoinKind.EXACT_TOKENS_IN_FOR_BPT_OUT) {
+      const [, amountsIn] =
+        WeightedPoolDecoder.joinExactTokensInForBPTOut(encodedUserData);
+      return amountsIn;
+    } else if (joinKind === WeightedPoolJoinKind.TOKEN_IN_FOR_EXACT_BPT_OUT) {
+      const [, bptAmountOut, tokenIndex] =
+        WeightedPoolDecoder.joinTokenInForExactBPTOut(encodedUserData);
+      return [bptAmountOut.toString(), tokenIndex];
+    } else throw new Error('Non supported join data');
+  }
+
+  allTokensInForExactBPTOut(encodedUserData: string, pool: Pool): string {
+    console.log(encodedUserData, pool);
+    throw new Error('joinAllTokensInForExactBPTOut not supported');
+    /*
+          We need maths for _calcAllTokensInGivenExactBptOut
+          From SC:
+          uint256 bptAmountOut = userData.allTokensInForExactBptOut();
+          // Note that there is no maximum amountsIn parameter: this is handled by `IVault.joinPool`.
+    
+          uint256[] memory amountsIn = WeightedMath._calcAllTokensInGivenExactBptOut(balances, bptAmountOut, totalSupply);
+    
+          return (bptAmountOut, amountsIn);
+    
+          const bptAmountOut = this.decodeJoinData(
+          encodedUserData,
+          PoolJoinKind.ALL_TOKENS_IN_FOR_EXACT_BPT_OUT
+        );
+        */
+  }
+
+  joinExactTokensInForBPTOut(
+    encodedUserData: string,
+    pool: Pool
+  ): [string, string[], string[]] {
+    // This does not include a value for pre-minted BPT
+    const amountsInWithRef = this.decodeJoinData(
+      encodedUserData,
+      WeightedPoolJoinKind.EXACT_TOKENS_IN_FOR_BPT_OUT
+    ) as string[];
+
+    const amountsIn =
+      this.relayerModel.doChainedRefReplacements(amountsInWithRef);
+    // Calculate amount of BPT out given exact amounts in
+    const bptAmountOut = pool._calcBptOutGivenExactTokensIn(
+      amountsIn.map((a) => BigNumber.from(a))
+    );
+
+    // Updates BPT/totalShares value for pool
+    if (
+      pool.SubgraphType == 'StablePhantom' ||
+      pool.SubgraphType == 'ComposableStable'
+    ) {
+      // Update BPT balance
+      // totalShares will be updated as a side effect within SOR
+      const bptAsPoolToken = pool.tokens.find((t) =>
+        isSameAddress(t.address, pool.address)
+      );
+      if (!bptAsPoolToken)
+        throw new Error('Pool does not contain BPT as a token');
+      const bptBalanceEVM = parseFixed(
+        bptAsPoolToken.balance.toString(),
+        bptAsPoolToken.decimals
+      );
+      pool.updateTokenBalanceForPool(
+        pool.address,
+        bptBalanceEVM.sub(bptAmountOut) // subtract value because joining pool reduces pre-minted BPT being held by the pool
+      );
+    } else {
+      // For pools that do not contain BPT as a token, update totalShares directly
+      pool.updateTokenBalanceForPool(
```

Comment:
@johngrantuk - just so I don't forget to bring this up, I had to revert this to `updateTokenBalanceForPool` instead of the newly created method `updateTotalShares` because the code expects the token balance to be updated as well. Updating total shares needs to be done as a side effect rather than being updated directly.
I'm not sure if there will be a scenario where we will want to update total shares directly without updating the BPT token balance, so maybe it would be best to remove `updateTotalShares` or at least make it a private method instead of exposing it on the SOR.